torch/fx/experimental/symbolic_shapes.py

```diff
@@ -1362,6 +1362,160 @@ def rbinary_magic_impl(self, other):
 del method
 del func
 
+try:
+    import z3
+
+    class SympyToZ3:
+        def __init__(self, symbols: Dict[sympy.Symbol, z3.BitVecRef]):
+            self._symbols = symbols
+
+        def _div(self, num, divisor):
+            return self._to_real(num) / self._to_real(divisor)
+
+        def _to_real(self, x):
+            return x if x.is_real() else z3.ToReal(x)
+
+        def _to_int(self, x):
+            return x if x.is_int() else z3.ToInt(x)
+
+        def _int(self, expr: int):
+            return z3.IntVal(expr)
+
+        def _Symbol(self, expr: sympy.Symbol):
+            return self._symbols[expr]
+
+        def _Rel(self, expr: sympy.Rel):
+            opmap = {
+                "==": operator.eq,
+                "!=": operator.ne,
+                ">=": operator.ge,
+                ">": operator.gt,
+                "<=": operator.le,
+                "<": operator.lt,
+            }
+            assert isinstance(expr, (sympy.Eq, sympy.Ne, sympy.Ge, sympy.Gt, sympy.Le, sympy.Lt))
+            assert expr.rel_op in opmap, (
+                f"sympy relop not supported: {expr.rel_op}"
+            )
+            return opmap[expr.rel_op](self(expr.lhs), self(expr.rhs))
+
+        def _Add(self, expr: sympy.Add):
+            return sum(self(expr.args))
+
+        def _Mod(self, expr: sympy.Mod):
+            p, q = [self._to_int(a) for a in self(expr.args)]
+            return p % q
+
+        def _Mul(self, expr: sympy.Mul):
+            numerator_list = []
+            denominator_list = []
+
+            for item in expr.args:
+                if item.is_commutative and isinstance(item, sympy.Pow) and item.exp.is_Rational and item.exp.is_negative:
+                    denominator_list.append(sympy.Pow(item.base, -item.exp))
+                else:
+                    numerator_list.append(item)
+
+            numerator = self._int(1) if len(numerator_list) == 0 else math.prod(self(numerator_list))
+
+            if len(denominator_list) > 0:
+                denominator = math.prod(self(denominator_list))
+                return self._div(numerator, denominator)
+
+            return numerator
+
+        def _Pow(self, expr: sympy.Pow):
+            base, exp = self(expr.args)
+            return base ** exp
+
+        def _FloorDiv(self, expr: FloorDiv):
+            base, divisor = self(expr.args)
+            return base / divisor
```

Comment:
I hope this is right, I couldn't easily tell from z3 docs lol

Reply:
This could be bad if base/divisor are real

Reply:
Right, this is wrong, even for ints (!).
I don't know why reals are needed for shape analysis, but I'll trust they are needed :)

For integers, can we switch to using bit-vectors instead? We need to specify a max bitwidth, but then division is defined the way we want (floor div always).

Reply:
The way to make this work for ints is to convert the result from Real to int.
e.g.:
```python
res = base / divisor
return self._to_int(res) if base.is_int() else res
```

Reply:
For reals, you can just convert to int and then back to real.

Reply:
Ah, right. That sounds like way easier than what I was doing.